components/builder-core/src/metrics.rs

```diff
@@ -0,0 +1,224 @@
+use std::thread;
+use statsd::Client;
+use hab_core::env;
+
+// Statsd Application name
+pub const APP_NAME: &'static str = "bldr";
+
+// Statsd Listener Address
+pub const STATS_ENV: &'static str = "HAB_STATS_ADDR";
+
+// Supported metrics
+#[derive(Debug, Clone)]
+pub enum Counter {
+    SearchPackages,
+}
+
+// Supported metrics
+#[derive(Debug, Clone)]
+pub enum Gauge {
+    PackageCount,
+}
+
+// Helper types
+#[derive(Debug, Clone, Copy)]
+enum MetricType {
+    Counter,
+    Gauge,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum MetricOperation {
+    Increment,
+    Decrement,
+    SetValue,
+}
+
+type MetricId = &'static str;
+type MetricValue = f64;
+type MetricTuple = (MetricType, MetricOperation, MetricId, Option<MetricValue>);
+
+trait Metric {
+    fn id(&self) -> &'static str;
+}
+
+// One-time initialization
+static mut SENDER: *const Sender<MetricTuple> = 0 as *const Sender<MetricTuple>;
+
+static INIT: Once = ONCE_INIT;
+
+fn sender() -> Sender<MetricTuple> {
+    unsafe {
+        INIT.call_once(|| {
+            SENDER = Box::into_raw(Box::new(init()));
+        });
+        (*SENDER).clone()
+    }
+}
+
+// init creates a worker thread ready to receive and process metric events,
+// and returns a channel for use by metric senders
+fn init() -> Sender<MetricTuple> {
+    let (tx, rx) = channel::<MetricTuple>();
+    let (rztx, rzrx) = sync_channel(0); // rendezvous channel
+
+    thread::Builder::new()
+        .name("metrics".to_string())
+        .spawn(move || receive(rztx, rx))
+        .expect("couldn't start metrics thread");
+
+    match rzrx.recv() {
+        Ok(()) => tx,
+        Err(e) => panic!("metrics thread startup error, err={}", e),
+    }
+}
+
+// receive runs in a separate thread and processes all metrics events
+fn receive(rz: SyncSender<()>, rx: Receiver<MetricTuple>) {
+    let mut client = statsd_client();
+    rz.send(()).unwrap(); // Blocks until the matching receive is called
+
+    loop {
+        let (mtyp, mop, mid, mval): MetricTuple = rx.recv().unwrap();
+        debug!("Received metrics tuple: {:?}", (mtyp, mop, mid, mval));
+
+        match client {
+            Some(ref mut cli) => {
+                match mtyp {
+                    MetricType::Counter => {
+                        match mop {
+                            MetricOperation::Increment => cli.incr(mid),
+                            MetricOperation::Decrement => cli.decr(mid),
+                            _ => error!("Unexpected metric operation: {:?}", mop),
+                        }
+                    }
+                    MetricType::Gauge => {
+                        match mop {
+                            MetricOperation::SetValue => cli.gauge(mid, mval.unwrap()),
+                            _ => error!("Unexpected metric operation: {:?}", mop),
+                        }
+                    }
+                }
+            }
+            None => (),
+        }
+    }
+}
+
+fn statsd_client() -> Option<Client> {
```

Comment:
I'd still prefer to see a `fn statsd_client(url: &Url)` or secondarily `fn statsd_client(url: &str)` as the mechanism by which to inject the Stasd service dependency. At the end of the day the Builder components will be using and consuming their own config (as well as setting suitable defaults for their `Config` structs) so we'll need that plumbing eventually. Additionally, if the string provided by the user/environment doesn't parse out to an appropriate URL I suspect that the `Client::new(...).unwrap()` call will `panic!` the metrics thread and will panic the main thread as a result (when the client doesn't properly start). Finally, having an environment variable drive this makes the service spawning more complicated as we need to add `Environment` declarations in our systemd units (or an extra environment variable if launched via a container) that is out-of-band from the remaining service config (all provided via Habitat).